packages/spectral/src/types/ConfigVars.ts

```diff
@@ -194,62 +181,49 @@ export type StandardConfigVar =
-> = TDataSourceType extends CollectionDataSourceType
-  ? {
-      dataSourceType: TDataSourceType;
-      collectionType?: CollectionType | undefined;
-    } & BaseConfigVar
-  : TDataSourceType extends Exclude<DataSourceType, CollectionDataSourceType>
-  ? BaseConfigVar & {
-      dataSourceType: TDataSourceType;
-      collectionType?: undefined;
-    }
-  :
-      | ({
-          dataSourceType: Extract<CollectionDataSourceType, TDataSourceType>;
-          collectionType: CollectionType;
-        } & BaseConfigVar)
-      | (BaseConfigVar & {
-          dataSourceType: Extract<
-            Exclude<DataSourceType, CollectionDataSourceType>,
-            TDataSourceType
-          >;
+type BaseDataSourceConfigVar<TDataSourceType extends DataSourceType = DataSourceType> =
+  TDataSourceType extends CollectionDataSourceType
+    ? {
+        dataSourceType: TDataSourceType;
+        collectionType?: CollectionType | undefined;
+      } & BaseConfigVar
+    : TDataSourceType extends Exclude<DataSourceType, CollectionDataSourceType>
+      ? BaseConfigVar & {
+          dataSourceType: TDataSourceType;
           collectionType?: undefined;
-        });
-
-type DataSourceDefinitionConfigVar =
-  DataSourceType extends infer TDataSourceType
-    ? TDataSourceType extends DataSourceType
-      ? BaseDataSourceConfigVar<TDataSourceType> &
-          Omit<
-            DataSourceDefinition<
-              Inputs,
-              ConfigVarResultCollection,
-              TDataSourceType
-            >,
-            "display" | "inputs" | "examplePayload" | "detailDataSource"
-          >
-      : never
-    : never;
+        }
+      :
+          | ({
+              dataSourceType: Extract<CollectionDataSourceType, TDataSourceType>;
+              collectionType: CollectionType;
+            } & BaseConfigVar)
+          | (BaseConfigVar & {
+              dataSourceType: Extract<
+                Exclude<DataSourceType, CollectionDataSourceType>,
+                TDataSourceType
+              >;
+              collectionType?: undefined;
+            });
+
+type DataSourceDefinitionConfigVar = DataSourceType extends infer TDataSourceType
+  ? TDataSourceType extends DataSourceType
+    ? BaseDataSourceConfigVar<TDataSourceType> &
+        Omit<
+          DataSourceDefinition<Inputs, ConfigVarResultCollection, TDataSourceType>,
+          "display" | "inputs" | "examplePayload" | "detailDataSource"
+        >
+    : never
+  : never;
 
 type DataSourceReferenceConfigVar =
-  ComponentRegistryDataSource extends infer TDataSourceReference
-    ? TDataSourceReference extends ComponentRegistryDataSource
-      ? Omit<
-          BaseDataSourceConfigVar<TDataSourceReference["dataSourceType"]>,
-          "dataSourceType"
-        > & {
-          dataSource: TDataSourceReference["reference"];
-        }
-      : never
+  ComponentRegistryDataSource extends infer TDataSourceReference extends ComponentRegistryDataSource
```

Comment:
This change alongside the bump past TS 5.1 is really the fix here. `MyType extends infer TMyType extends MyType` is narrower than `MyType extends infer TMyType ? TMyType extends MyType ? ...`

Reply:
Looks way better